### PR TITLE
Allow environment views to be sym/hard link and copy types

### PR DESCRIPTION
**Pull request by psakievich**
Add a syntax to specify view type as copy/symlink/hardlink.
This can stand with or without the commit by @vsoch which is in #24087.

### CODE REVIEW

**Comment by psakievich** [comment]
This is also not really working yet.  I created the following environment:
```
spack:
  specs: [pkgconf]
  view:
    copy:
      root: /Users/psakiev/soft/test/copy
      link_type: copy
    symlink:
      root: /Users/psakiev/soft/test/symlink
      link_type: symlink
    hardlink:
      root: /Users/psakiev/soft/test/hardlink
      link_type: hardlink
```
But I need to resolve an error

~~I haven't really dug into the logic of how the environment view's create the view in the hidden directory and then symlink to that. I think some changes will have to be made there.  @becker33 and @tgamblin the git history looks like you are the primary authors of that section.~~  I'd also welcome any suggestions on how to unit test this better to differentiate between the different link types.

Update: Issue was not related to this, but rather due to function argument type issues.

**Comment by psakievich** [comment]
My test from above is working now.  Looks like it was a string vs function in the parsing that I missed.

**Comment by psakievich** [comment]
@spackbot run pipeline

**Comment by alalazo** [comment]
```
Error: Pipeline generation failed due to the presence of the following specs that are known to be broken in develop:
* omega-h@9.32.5%gcc@7.5.0~examples~ipo+mpi+optimize+shared+symbols~throw+trilinos~warnings+zlib build_type=RelWithDebInfo arch=linux-ubuntu18.04-x86_64 ^berkeley-db@18.1.40%gcc@7.5.0+cxx~docs+stl patches=b231fcc4d5cff05e5c3a4814f6a5af0e9a966428dc2176540d2c05aff41de522 arch=linux-ubuntu18.04-x86_64 ^bzip2@1.0.8%gcc@7.5.0~debug~pic+shared arch=linux-ubuntu18.04-x86_64 ^cmake@3.20.3%gcc@7.5.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-ubuntu18.04-x86_64 ^diffutils@3.7%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^findutils@4.8.0%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^gdbm@1.19%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^hwloc@2.4.1%gcc@7.5.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml+pci+shared arch=linux-ubuntu18.04-x86_64 ^libfabric@1.12.1%gcc@7.5.0~debug~kdreg fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu18.04-x86_64 ^libiconv@1.16%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^libpciaccess@0.16%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^libsigsegv@2.13%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^libtool@2.4.6%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^libxml2@2.9.10%gcc@7.5.0~python arch=linux-ubuntu18.04-x86_64 ^m4@1.4.19%gcc@7.5.0+sigsegv arch=linux-ubuntu18.04-x86_64 ^mpich@3.4.2%gcc@7.5.0~argobots+fortran+hwloc+hydra+libxml2+pci+romio~slurm~verbs~wrapperrpath device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu18.04-x86_64 ^ncurses@6.2%gcc@7.5.0~symlinks+termlib abi=none arch=linux-ubuntu18.04-x86_64 ^openblas@0.3.15%gcc@7.5.0~bignuma~consistent_fpcsr~ilp64+locking+pic+shared threads=openmp arch=linux-ubuntu18.04-x86_64 ^openssl@1.1.1k%gcc@7.5.0~docs+systemcerts arch=linux-ubuntu18.04-x86_64 ^perl@5.34.0%gcc@7.5.0+cpanm+shared+threads arch=linux-ubuntu18.04-x86_64 ^pkgconf@1.7.4%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^readline@8.1%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^trilinos@13.0.1%gcc@7.5.0~adios2+amesos+amesos2+anasazi+aztec~basker+belos~boost~cgns~chaco~complex~cuda~cuda_rdc~debug~dtk+epetra+epetraext~epetraextbtf~epetraextexperimental~epetraextgraphreorderings~exodus+explicit_template_instantiation~float+fortran~hdf5~hypre+ifpack+ifpack2~intrepid~intrepid2~ipo~isorropia+kokkos~matio~mesquite~minitensor+ml+mpi+muelu~mumps~netcdf~nox~openmp~phalanx~piro~pnetcdf~python~rol~rythmos+sacado~scorec~shards+shared~shylu~stk~stokhos~stratimikos~strumpack~suite-sparse~superlu~superlu-dist~teko~tempus+tpetra~trilinoscouplings~wrapper~zlib~zoltan~zoltan2 build_type=RelWithDebInfo cuda_arch=none cxxstd=11 gotype=long arch=linux-ubuntu18.04-x86_64 ^util-macros@1.19.3%gcc@7.5.0 arch=linux-ubuntu18.04-x86_64 ^xz@5.2.5%gcc@7.5.0+pic libs=shared,static arch=linux-ubuntu18.04-x86_64 ^zlib@1.2.11%gcc@7.5.0+optimize+pic+shared arch=linux-ubuntu18.04-x86_64 (ez3q5hbipkxqvscxc27ej3rxowrb4wfq)
```

**Comment by psakievich** [comment]
@alalazo these failures don't have anything to do with my PR correct?

**Comment by tgamblin** [comment]
@zackgalbreath @scottwittenburg when you get a chance can you look at this pipeline?

If this PR requires `omega-h` to rebuild somehow, and if it gets the same full hash as a `develop` build that failed, the pipeline will fail b/c it considers that build to be broken.  I am not seeing anything here that would cause any of our pipeline packages to *need* to rebuild, so I'm wondering if this is a false positive.

**Comment by psakievich** [comment]
Ah thanks for the explanation @tgamblin.

**Comment by scottwittenburg** [comment]
I just looked at all the gitlab pipelines that ran on this branch, the most recent one passed merged with a `develop` from yesterday.  The previous one failed because it generated a job with the hash of the broken `omega-h`, and that was against a `develop` from about a week ago.  Since we're not testing every github merge commit that comes along, this seems normal to me.  What I don't understand is how that successful pipeline from yesterday got started with the newer merge commit.  For that to happen, I would have expected to see a change pushed sometime after the spackbot-triggered pipeline from two days ago (which doesn't update the merge commit, but just tests whatever current commit is on the branch) failed.

**Comment by psakievich** [comment]
@tgamblin @becker33 @vsoch bump for review and merge?

File: lib/spack/spack/filesystem_view.py [hunk 1]
```diff
@@ -98,6 +98,28 @@ def view_copy(src, dst, view, spec=None):
...
 
+def view_func_parser(parsed_name):
+    # What method are we using for this view
+    if parsed_name in ("hardlink", "hard"):
+        link_fn = view_hardlink
+    elif parsed_name in ("copy", "relocate"):
+        link_fn = view_copy
+    else:
+        link_fn = view_symlink
+    return link_fn
```
Comment (tgamblin) [inline lib/spack/spack/filesystem_view.py]: This should check for `'symlink'` and raise a `ValueError` for unknown values:

```suggestion
    if parsed_name in ("hardlink", "hard"):
        return view_hardlink
    elif parsed_name in ("copy", "relocate"):
        return view_copy
    elif parsed_name == "symlink":
        return view_symlink
    else:
        raise ValueError("invalid link type for view: '%s'" % parsed_name)
```

File: lib/spack/spack/test/cmd/env.py [hunk 2]
```diff
@@ -1948,6 +1948,48 @@ def test_view_link_roots(tmpdir, mock_fetch, mock_packages, mock_archive,
...
+        for spec in test._get_environment_specs():
+            if spec in test.roots() and (spec.satisfies('%gcc') and
+                                         not spec.satisfies('callpath')):
+                assert os.path.exists(
+                    os.path.join(viewdir, spec.name, '%s-%s' %
+                                 (spec.version, spec.compiler.name)))
+            else:
+                assert not os.path.exists(
+                    os.path.join(viewdir, spec.name, '%s-%s' %
+                                 (spec.version, spec.compiler.name)))
```
Comment (tgamblin) [inline lib/spack/spack/test/cmd/env.py]: Shouldn't this test assert that the files are actually copied in addition to the fact that they exist?

I would verify `not os.path.islink()` for these -- to ensure that it's not just doing normal symlinking.

**Review by tgamblin — DISMISSED** [review]

File: lib/spack/spack/test/cmd/env.py (hunk 2 above)
Comment (psakievich) [reply lib/spack/spack/test/cmd/env.py]: Yeah the link test is a good idea. In terms of testing that files are actually copied, do the mocks actually generate files to check inside the install directories?  I started diving into that but didn't get very far.

File: lib/spack/spack/filesystem_view.py [hunk 3]
```diff
@@ -98,6 +98,29 @@ def view_copy(src, dst, view, spec=None):
             )
 
 
+def view_func_parser(parsed_name):
+    # What method are we using for this view
+    if parsed_name in ("hardlink", "hard"):
+        return view_hardlink
+    elif parsed_name in ("copy", "relocate"):
+        return view_copy
+    elif parsed_name == "symlink":
```
Comment (becker33) [inline lib/spack/spack/filesystem_view.py]:
```suggestion
    elif parsed_name in ("add", "symlink", "soft"):
```

This should match the options from the command:
```
$ spack view -h
usage: spack view [-hv] [-e EXCLUDE] [-d {true,false,yes,no}] ACTION ...

project packages to a compact naming scheme on the filesystem.

positional arguments:
  ACTION
    symlink (add, soft)
                        add package files to a filesystem view via symbolic links
    hardlink (hard)     add packages files to a filesystem view via hard links
    copy (relocate)     add package files to a filesystem view via copy/relocate
    remove (rm)         remove packages from a filesystem view
    statlink (status, check)
                        check status of packages in a filesystem view

optional arguments:
  -h, --help            show this help message and exit
  -v, --verbose         If not verbose only warnings/errors will be printed.
  -e EXCLUDE, --exclude EXCLUDE
                        exclude packages with names matching the given regex pattern
  -d {true,false,yes,no}, --dependencies {true,false,yes,no}
                        Link/remove/list dependencies.
```

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/cmd/env.py (hunk 2 above)
Comment (psakievich) [reply lib/spack/spack/test/cmd/env.py]: The tricky thing here is that the first file location you'd test is a link. For example in my test environment above the top directories `copy`, `hardlink` and `symlink` are all links. The distinction is the files inside `copy/bin` aren't links and the files inside `symlink/bin` are.  I've tried adding:
```
        test = ev.read('test')
        for spec in test._get_environment_specs():
            if spec in test.roots() and (spec.satisfies('%gcc') and
                                         not spec.satisfies('callpath')):
                assert os.path.exists(
                    os.path.join(viewdir, spec.name, '%s-%s' %
                                 (spec.version, spec.compiler.name), 'bin'))
                assert not os.path.islink(
                    os.path.join(viewdir, spec.name, '%s-%s' %
                                 (spec.version, spec.compiler.name), 'bin')
```
but it looks like the mock's aren't creating the `bin` directory.

File: lib/spack/spack/filesystem_view.py (hunk 3 above)
Comment (tgamblin) [reply lib/spack/spack/filesystem_view.py]: must we have so many aliases?  Like "add"  Does anyone use add?

Comment (tgamblin) [reply lib/spack/spack/filesystem_view.py]: Seems like we could deprecate some of these.

File: lib/spack/spack/test/cmd/env.py (hunk 2 above)
Comment (psakievich) [reply lib/spack/spack/test/cmd/env.py]: Pretty sure I got this figured out with @becker33's help on slack.  I added two more unit tests to check all three link types.

File: lib/spack/spack/test/cmd/env.py [hunk 4]
```diff
@@ -1948,6 +1948,154 @@ def test_view_link_roots(tmpdir, mock_fetch, mock_packages, mock_archive,
                                  (spec.version, spec.compiler.name)))
 
 
+def test_view_link_type_symlink(tmpdir, mock_fetch, mock_packages, mock_archive,
```
Comment (becker33) [inline lib/spack/spack/test/cmd/env.py]: We could make all of these tests simpler by:

1. Removing the definitions from the environment
2. Setting the environment specs to `[mpileaks]`
3. Removing the projections, select, exclude, and link keys from the view descriptor
4. Adjusting the paths we check for appropriately

We should also be able to parametrize the three tests into one.

```
@pytest.mark.parametrize('type', ['symlink', 'hardlink', 'copy'])
def test_view_link_type(type, tmpdir, mock_fetch, mock_packages, mock_archive,
...
...
assert os.path.islink(...) == (type == 'symlink')
```

**Review by becker33 — CHANGES_REQUESTED** [review]
The content all looks good, just had an idea how we can simplify the new tests to be more maintainable.

**Review by becker33 — APPROVED** [review]

File: lib/spack/spack/test/cmd/env.py [hunk 5]
```diff
@@ -1948,6 +1948,37 @@ def test_view_link_roots(tmpdir, mock_fetch, mock_packages, mock_archive,
...
+        test = ev.read('test')
+
+        for spec in test._get_environment_specs():
+            view_path = os.path.join(viewdir, spec.name)
+            assert os.path.exists(view_path)
+
+            file_to_test = os.path.join(view_path, spec.name)
+
+            assert os.path.isfile(file_to_test)
+            assert os.path.islink(file_to_test) == (link_type == 'symlink')
```
Comment (becker33) [inline lib/spack/spack/test/cmd/env.py]:
```suggestion
            file_to_test = os.path.join(test.default_view.view()._root, spec.name)

            assert os.path.isfile(file_to_test)
            assert os.path.islink(file_to_test) == (link_type == 'symlink')
```

**Review by becker33 — CHANGES_REQUESTED** [review]
I think this is what you need to test this effectively.

**Review by becker33 — APPROVED** [review]
I think this is what you need to test this effectively.

**Comment by psakievich** [comment]
@spackbot run pipeline

**Comment by becker33** [comment]
@psakievich I've merged this, but we do still need to add documentation for it.

**Comment by vsoch** [comment]
Do the little bit of docs here: https://github.com/spack/spack/pull/24832#issuecomment-878269972 still reflect how to use the new feature? I'd like to give it a shot for the spack / snakemake integration I was testing.

**Comment by psakievich** [comment]
@vsoch yes that is it.

**Comment by haampie** [comment]
Nice, this is a useful feature, thanks!

**Comment by vsoch** [comment]
See https://github.com/spack/spack/pull/25465 for a first shot at docs for this. I think we will want to update this with what exactly the root means (I'm testing it out, and I put the root that I created for the view and it's getting angry at me).

**Comment by psakievich** [comment]
@vsoch not sure I follow this:
>   I put the root that I created for the view and it's getting angry at me

Could you elaborate more regarding this anger?

**Comment by vsoch** [comment]
haha yes certainly! So the script creates that root directory first and then plops the spack.yaml in it, and runs the env create command:

```bash
Creating spack environment envs/spack-stats.yml...
spack env create --dir '/home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2' /home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2/spack.yaml
```
And note that the root is that same path, because that's where I want the environment to be created (and where it would be created if I didn't add "view"

```
/home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2' /home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2
```
And here is the anger, lol. The top line is a generic snakemake error that caught the command error, below it is the error from spack about not being able to create the view:

```bash
snakemake.exceptions.CreateSpackEnvironmentException: Could not create spack environment from /home/vanessa/Desktop/Code/snakemake/tutorial/envs/spack-stats.yml:
==> Error: Cannot create view: file already exists and is not a link: /home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2
```

**Comment by vsoch** [comment]
So the spack.yaml at one of the hashes: /home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2/spack.yaml looks like this:

```yaml
spack:
  concretization: together
  specs:
  - py-pysam
  - py-matplotlib
  - python@3.8.8
  view:
    copy:
      link_type: copy
      root: /home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2
```
Should I be putting something else under root, or doing create differently?

**Comment by psakievich** [comment]
@vsoch if my hash reading dyslexia is not mistaken it looks like you are giving it the `$SPACK_ENV` directory. The `root` directory in this scope is only for the view.   If you omit it then spack will create a directory inside `/home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2/` where the view will be created.  

You can't give the view `root` a pre-existing directory because spack actually symlinks this `root` directory for the view to a hidden directory.  This is done so it can update views without changing the root directory. This is why you are getting that error.

So to get this to work I would just omit the `root` specification, or set it to something like:
`root: /home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2/my_spiffy_view`

This specified `root` will then be linked against a directory 
`/home/vanessa/Desktop/Code/snakemake/tutorial/.snakemake/spack/827d39341264ddaff183ebb496b21eb2/._my_spiffy_view`

**Comment by vsoch** [comment]
The first didn't work (removing it entirely) - but the second is running the spack install now!

**Comment by psakievich** [comment]
huh I wonder what went wrong with removing it entirely.  I will try to replicate this. None the less, glad you're getting a spack install!

**Pull request merged**